### PR TITLE
feat(common): return an `InvalidRequest` error when a parameter is invalid - 6.3.20

### PR DESCRIPTION
**Pull request by thomasBousselin**
Still need to add the annotation to all the endpoints.

### CODE REVIEW

File: search-service/src/main/kotlin/com/egm/stellio/search/entity/web/EntityHandler.kt [hunk 1]
```diff
@@ -323,11 +396,17 @@ class EntityHandler(
     suspend fun appendEntityAttributes(
         @RequestHeader httpHeaders: HttpHeaders,
         @PathVariable entityId: URI,
-        @RequestParam options: Optional<String>,
-        @RequestBody requestBody: Mono<String>
+        @RequestBody requestBody: Mono<String>,
+        @RequestParam
+        @AllowedParameters(
+            implemented = [QueryParameter.OPTIONS], // todo type no implemented?
```
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/entity/web/EntityHandler.kt:402]: `type` is for distOps

File: search-service/src/main/kotlin/com/egm/stellio/search/temporal/util/TemporalQueryParamsUtils.kt [hunk 2]
```diff
@@ -1,10 +1 @@
 package com.egm.stellio.search.temporal.util
-
-const val TIMEREL_PARAM = "timerel"
-const val TIMEAT_PARAM = "timeAt"
-const val ENDTIMEAT_PARAM = "endTimeAt"
-const val AGGRPERIODDURATION_PARAM = "aggrPeriodDuration"
-const val AGGRMETHODS_PARAM = "aggrMethods"
-const val LASTN_PARAM = "lastN"
-const val TIMEPROPERTY_PARAM = "timeproperty"
-const val WHOLE_TIME_RANGE_DURATION = "PT0S"
```
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/temporal/util/TemporalQueryParamsUtils.kt:10]: `WHOLE_TIME_RANGE_DURATION` is not a query param (just a special value we have to take care of when calculating aggregations, more like an implementation detail)

File: shared/src/main/kotlin/com/egm/stellio/shared/model/LinkedEntityQuery.kt [hunk 3]
```diff
@@ -10,16 +15,47 @@ data class LinkedEntityQuery(
 ) {
     companion object {
         const val DEFAULT_JOIN_LEVEL = 1
-    }
 
-    enum class JoinType(val type: String) {
```
Comment (bobeal) [inline shared/src/main/kotlin/com/egm/stellio/shared/model/LinkedEntityQuery.kt:15]: why did you move it out of the data class?

File: shared/src/main/kotlin/com/egm/stellio/shared/model/parameter/AllowedParameters.kt [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        ConstraintValidator<AllowedParameters, MultiValueMap<String, String>?> {
+        private var implemented: List<String> = listOf()
+        private var notImplemented: List<String> = listOf()
+
+        override fun initialize(requiredIfChecked: AllowedParameters) {
+            this.implemented = requiredIfChecked.implemented.map(QueryParameter::key)
+            this.notImplemented = requiredIfChecked.notImplemented.map(QueryParameter::key)
+        }
+
+        override fun isValid(value: MultiValueMap<String, String>?, context: ConstraintValidatorContext): Boolean {
```
Comment (bobeal) [inline shared/src/main/kotlin/com/egm/stellio/shared/model/parameter/AllowedParameters.kt:34]: can `value` really be null?

File: shared/src/main/kotlin/com/egm/stellio/shared/model/parameter/AllowedParameters.kt [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    val message: String = "Invalid parameter received",
+    val groups: Array<KClass<*>> = [],
+    val payload: Array<KClass<*>> = [],
+) {
+    class ParamValidator :
+        ConstraintValidator<AllowedParameters, MultiValueMap<String, String>?> {
+        private var implemented: List<String> = listOf()
+        private var notImplemented: List<String> = listOf()
+
+        override fun initialize(requiredIfChecked: AllowedParameters) {
```
Comment (bobeal) [inline shared/src/main/kotlin/com/egm/stellio/shared/model/parameter/AllowedParameters.kt:29]: don't really get what means this name: `requiredIfChecked`?

File: shared/src/main/kotlin/com/egm/stellio/shared/model/parameter/AllowedParameters.kt [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+                "Accepted parameters are '${implemented.joinToString("', '")}'"
+            ).addConstraintViolation()
+
+            if (value.keys.any { it in notImplemented }) {
+                context.buildConstraintViolationWithTemplate(
+                    "The '${notImplemented.joinToString("', '")}' parameters are not implemented yet "
+                ).addPropertyNode(HttpStatus.NOT_IMPLEMENTED.name).addConstraintViolation()
+                return false
+            }
+            return false
```
Comment (bobeal) [inline shared/src/main/kotlin/com/egm/stellio/shared/model/parameter/AllowedParameters.kt:49]: it would also be nice to say which parameter(s) has(ve) caused the error?

File: shared/src/main/kotlin/com/egm/stellio/shared/model/parameter/GeoQuery.kt [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+import com.egm.stellio.shared.model.stringifyCoordinates
+import com.egm.stellio.shared.util.JsonLdUtils
+import com.egm.stellio.shared.util.JsonLdUtils.JSONLD_VALUE
+import com.egm.stellio.shared.util.JsonLdUtils.NGSILD_GEOPROPERTY_VALUE
+import com.egm.stellio.shared.util.JsonLdUtils.NGSILD_LOCATION_PROPERTY
+import com.egm.stellio.shared.util.JsonLdUtils.expandJsonLdTerm
+import com.egm.stellio.shared.util.JsonUtils
+import com.egm.stellio.shared.util.decode
+
+data class GeoQuery(
```
Comment (bobeal) [inline shared/src/main/kotlin/com/egm/stellio/shared/model/parameter/GeoQuery.kt:22]: not clear to me why `GeoQuery` has been moved in this package but not `LinkedEntityQuery`, `PaginationQuery`,...

File: shared/src/main/kotlin/com/egm/stellio/shared/model/GeoQueryUtils.kt [hunk 8]
```diff
@@ -0,0 +1,18 @@
+package com.egm.stellio.shared.model
```
Comment (bobeal) [inline shared/src/main/kotlin/com/egm/stellio/shared/model/GeoQueryUtils.kt:1]: don't get why what remains from `util/GeoQueryUtils` is now in `model` (actually, its place is in `util`)

File: shared/src/main/kotlin/com/egm/stellio/shared/web/ExceptionHandler.kt [hunk 9]
```diff
@@ -35,6 +38,16 @@ class ExceptionHandler {
                 NotAcceptableException(cause.message).toErrorResponse()
             is MethodNotAllowedException ->
                 ResponseEntity.status(HttpStatus.METHOD_NOT_ALLOWED).body(cause.body)
+            is ConstraintViolationException -> {
+                val message = cause.constraintViolations.map { it.message }.joinToString(" . ")
```
Comment (bobeal) [inline shared/src/main/kotlin/com/egm/stellio/shared/web/ExceptionHandler.kt:42]:
```suggestion
                val message = cause.constraintViolations.joinToString(". ") { it.message }

```

File: shared/src/main/kotlin/com/egm/stellio/shared/web/ExceptionHandler.kt [hunk 10]
```diff
@@ -35,6 +38,16 @@ class ExceptionHandler {
...
+                if (cause.constraintViolations.any // todo simplify
+                        { constraint -> constraint.propertyPath.any { it.name == HttpStatus.NOT_IMPLEMENTED.name } }
+                ) {
```
Comment (bobeal) [inline shared/src/main/kotlin/com/egm/stellio/shared/web/ExceptionHandler.kt:45]: don't really need the `{` here

File: shared/src/main/resources/shared.properties [hunk 11]
```diff
@@ -8,7 +8,7 @@ application.authentication.enabled = false
 # Pagination config for query resources endpoints
 application.pagination.limit-default = 30
 application.pagination.limit-max = 100
-application.pagination.temporal-limit = 10000
+application.pagination.temporal-limit = 10
```
Comment (bobeal) [inline shared/src/main/resources/shared.properties:11]: to be removed

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/job/TimeIntervalNotificationJob.kt [hunk 12]
```diff
@@ -57,12 +53,14 @@ class TimeIntervalNotificationJob(
...
         val param = java.lang.StringBuilder()
-        param.append("?$QUERY_PARAM_TYPE=${entitySelector.typeSelection.encode()}")
-        if (entitySelector.id != null) param.append("&$QUERY_PARAM_ID=${entitySelector.id}")
-        if (entitySelector.idPattern != null) param.append("&$QUERY_PARAM_ID_PATTERN=${entitySelector.idPattern}")
-        if (q != null) param.append("&$QUERY_PARAM_Q=${q.encode()}")
+        param.append("?${QueryParameter.TYPE.key}=${entitySelector.typeSelection.encode()}")
+        if (entitySelector.id != null) param.append("&${QueryParameter.ID.key}=${entitySelector.id}")
+        if (entitySelector.idPattern != null) param.append(
+            "&${QueryParameter.ID_PATTERN.key}=${entitySelector.idPattern}"
+        )
```
Comment (bobeal) [inline subscription-service/src/main/kotlin/com/egm/stellio/subscription/job/TimeIntervalNotificationJob.kt:60]:
```suggestion
        if (entitySelector.idPattern != null)
            param.append("&${QueryParameter.ID_PATTERN.key}=${entitySelector.idPattern}")
```

File: shared/src/main/kotlin/com/egm/stellio/shared/model/parameter/GeoQuery.kt (hunk 7 above)
Comment (bobeal) [reply shared/src/main/kotlin/com/egm/stellio/shared/model/parameter/GeoQuery.kt:22]: more globally, this new `parameter` sub-package is a bit fuzzy.

File: shared/src/main/kotlin/com/egm/stellio/shared/model/parameter/QueryParameter.kt [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+
+    override fun toString(): String {
+        return key
+    }
+
+    object Query {
+        val qPattern: Pattern = Pattern.compile("([^();|]+)")
+        val typeSelectionRegex: Regex = """([^(),;|]+)""".toRegex()
+        val scopeSelectionRegex: Regex = """([^(),;|]+)""".toRegex()
+    }
```
Comment (bobeal) [inline shared/src/main/kotlin/com/egm/stellio/shared/model/parameter/QueryParameter.kt:70]: don't get what this is doing here

File: shared/src/main/kotlin/com/egm/stellio/shared/model/parameter/GeoQuery.kt [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+                entries.any { it.type == type }
+
+            fun forType(type: String): GeometryType? =
+                entries.find { it.type == type }
+        }
+    }
+
+    companion object {
+
+        fun parseGeoQueryParameters(
```
[the remark this answers is not in the document]
Comment (thomasBousselin) [reply shared/src/main/kotlin/com/egm/stellio/shared/model/parameter/GeoQuery.kt:48]: Yes, also it returns a Either so it can't really be in a constructor.

[the remark this answers is not in the document]
Comment (bobeal) [reply shared/src/main/kotlin/com/egm/stellio/shared/model/parameter/GeoQuery.kt:48]: It can be a constructor and return an either (it is done in some places)

File: search-service/src/main/kotlin/com/egm/stellio/search/temporal/util/TemporalQueryParamsUtils.kt (hunk 2 above)
Comment (thomasBousselin) [reply search-service/src/main/kotlin/com/egm/stellio/search/temporal/util/TemporalQueryParamsUtils.kt:10]: Yes i put it in TemporalQueryUtils instead. (because it is only used by search service)

File: search-service/src/main/kotlin/com/egm/stellio/search/entity/web/EntityOperationHandler.kt [hunk 15]
```diff
@@ -98,7 +103,12 @@ class EntityOperationHandler(
     suspend fun upsert(
         @RequestHeader httpHeaders: HttpHeaders,
         @RequestBody requestBody: Mono<String>,
-        @RequestParam(required = false) options: String?
+        @RequestParam(required = false) options: String?,
```
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/entity/web/EntityOperationHandler.kt:106]: a bit strange to have the `options` query parameter separated. for consistency, it should now be retrieve from `queryParams` (to be also checked in some other places)

File: search-service/src/main/kotlin/com/egm/stellio/search/csr/web/ContextSourceRegistrationHandler.kt [hunk 16]
```diff
@@ -144,7 +157,11 @@ class ContextSourceRegistrationHandler(
      * Implements 6.9.3.3 - Delete ContextSourceRegistration
      */
     @DeleteMapping("/{contextSourceRegistrationId}")
-    suspend fun delete(@PathVariable contextSourceRegistrationId: URI): ResponseEntity<*> = either {
+    suspend fun delete(
+        @PathVariable contextSourceRegistrationId: URI,
+        @AllowedParameters(notImplemented = [])
```
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/csr/web/ContextSourceRegistrationHandler.kt:162]: can't just do `@AllowedParameters`?

File: search-service/src/main/kotlin/com/egm/stellio/search/temporal/web/TemporalEntityOperationsHandler.kt [hunk 17]
```diff
@@ -38,6 +40,13 @@ class TemporalEntityOperationsHandler(
     suspend fun queryEntitiesViaPost(
         @RequestHeader httpHeaders: HttpHeaders,
         @RequestBody requestBody: Mono<String>,
+        @AllowedParameters( // todo double check GEOMETRY_PROPERTY, OPTIONS  in spec
```
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/temporal/web/TemporalEntityOperationsHandler.kt:43]: most of them are passed in the body (see GeoQuery - 5.2.13 and TemporalQuery - 5.2.21)

File: shared/src/main/kotlin/com/egm/stellio/shared/queryparameter/AllowedParameters.kt [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+
+            val notImplementedKeys = params.keys.intersect(notImplemented)
+            val errorKeys = params.keys - notImplementedKeys - implemented
+
+            context.disableDefaultConstraintViolation()
+
+            val message = StringBuilder().apply {
+                if (notImplementedKeys.isNotEmpty()) {
+                    append(
+                        "The '${notImplementedKeys.joinToString("', '")}' parameter is not implemented yet. "
```
Comment (bobeal) [inline shared/src/main/kotlin/com/egm/stellio/shared/queryparameter/AllowedParameters.kt:47]: in plural maybe? can we write it like this in English: `The ... parameter(s) is(are) not implemented yet`?

File: shared/src/main/kotlin/com/egm/stellio/shared/queryparameter/OptionsValue.kt [hunk 19]
```diff
@@ -0,0 +1,8 @@
+package com.egm.stellio.shared.queryparameter
+
+enum class OptionsValue(val value: String) {
+    SYS_ATTRS("sysAttrs"),
+    KEY_VALUES("keyValues"),
+    NO_OVERWRITE("noOverwrite"),
+    OBSERVED_AT("observedAt")
```
Comment (bobeal) [inline shared/src/main/kotlin/com/egm/stellio/shared/queryparameter/OptionsValue.kt:7]: `observatedAt` is a specific query parameter (and not a possible value for `options`)

**Comment by bobeal** [comment]
- pushed somes fixes and renamings
- it is missing tests (here and in the conformance test suite)

File: search-service/src/main/kotlin/com/egm/stellio/search/temporal/web/TemporalApiResponses.kt [hunk 20]
```diff
@@ -32,8 +32,10 @@ object TemporalApiResponses {
         lang: String? = null,
     ): ResponseEntity<String> {
         val baseRepresentation = parseRepresentations(requestParams, mediaType)
-
-        val representation = lang?.let { baseRepresentation.copy(languageFilter = it) } ?: baseRepresentation
+        val representation = lang?.let {
+            baseRepresentation.copy(languageFilter = it, timeproperty = query.temporalQuery.timeproperty.propertyName)
+        }
+            ?: baseRepresentation.copy(timeproperty = query.temporalQuery.timeproperty.propertyName)
```
[the remark this answers is not in the document]
Comment (thomasBousselin) [reply search-service/src/main/kotlin/com/egm/stellio/search/temporal/web/TemporalApiResponses.kt:38]: Yes, but it is set from the queryParameters, so this will fail for queryEntitiesViaPost.
It may not be the best way to do it. but it has the advantage of not needed to change the parseRepresentation function.

[the remark this answers is not in the document]
Comment (bobeal) [reply search-service/src/main/kotlin/com/egm/stellio/search/temporal/web/TemporalApiResponses.kt:38]: hmm, yes. I guess this is why I did the same for languageFilter...

can you just add a one line comment to explain the why?

File: shared/src/main/kotlin/com/egm/stellio/shared/queryparameter/AllowedParameters.kt [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+            val invalidKeys = params.keys - notImplementedKeys - implemented
+
+            context.disableDefaultConstraintViolation()
+
+            val message = StringBuilder().apply {
+                if (notImplementedKeys.isNotEmpty()) {
+                    append(
+                        "The '${
+                            notImplementedKeys.joinToString("', '")
+                        }' parameter(s) has/have not been implemented yet. "
```
[the remark this answers is not in the document]
Comment (thomasBousselin) [reply shared/src/main/kotlin/com/egm/stellio/shared/queryparameter/AllowedParameters.kt]: i asked gpt 😞

[the remark this answers is not in the document]
Comment (bobeal) [reply shared/src/main/kotlin/com/egm/stellio/shared/queryparameter/AllowedParameters.kt]: I was not talking about the message but about this thing on 3 lines:

```
'${
       notImplementedKeys.joinToString("', '")
}'
```

File: search-service/src/main/kotlin/com/egm/stellio/search/authorization/web/EntityAccessControlHandler.kt [hunk 22]
```diff
@@ -56,7 +56,7 @@ import reactor.core.publisher.Mono
 import java.net.URI
 
 @RestController
-@RequestMapping("/ngsi-ld/v1/entityAccessControl")
+@RequestMapping("/ngsi-ld/v1/entityAccessControl") // todo
```
[the remark this answers is not in the document]
Comment (thomasBousselin) [reply search-service/src/main/kotlin/com/egm/stellio/search/authorization/web/EntityAccessControlHandler.kt]: The "/entities","/users","/groups" look like the getEntities endpoint so i deduced that they support the same parameters.

[the remark this answers is not in the document]
Comment (bobeal) [reply search-service/src/main/kotlin/com/egm/stellio/search/authorization/web/EntityAccessControlHandler.kt]: there is nothing to deduce: https://stellio.readthedocs.io/en/latest/user/authentication_and_authorization.html

[the remark this answers is not in the document]
Comment (thomasBousselin) [reply search-service/src/main/kotlin/com/egm/stellio/search/authorization/web/EntityAccessControlHandler.kt]: contrary to what is told in the documentation i think the endpoint implement at least the pagination parameters. 
So should i reduce it to QP.ID, QP.TYPE, QP.ATTRS, QP.COUNT, QP.OFFSET, QP.LIMIT?
Or i remove the @AllowedParameters from those endpoint.

**Review by ranim-n — APPROVED** [review]

File: search-service/src/main/kotlin/com/egm/stellio/search/authorization/web/EntityAccessControlHandler.kt [hunk 23]
```diff
@@ -1,5 +1,6 @@
 package com.egm.stellio.search.authorization.web
```
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/authorization/web/EntityAccessControlHandler.kt:1]: miss the `@Validated` on the handler class
